artiq/compiler/validators/escape.py

```diff
@@ -301,17 +300,20 @@ def visit_ClassDefT(self, node):
     def visit_assignment(self, target, value):
         value_region  = self._region_of(value)
 
-        # If this is a variable, we might need to contract the live range.
-        if isinstance(value_region, Region):
-            for name in self._names_of(target):
-                region = self._region_of(name)
-                if isinstance(region, Region):
-                    region.contract(value_region)
-
         # If we assign to an attribute of a quoted value, there will be no names
         # in the assignment lhs.
         target_names   = self._names_of(target) or []
 
+        # Adopt the value region for any variables declared on the lhs.
+        for name in target_names:
+            region = self._region_of(name)
+            if isinstance(region, Region) and not region.present():
+                # Find the name's environment to overwrite the region.
+                for env in self.env_stack[::-1]:
+                    if name.id in env:
+                        env[name.id] = value_region
+                        break
```

Comment:
I'm not convinced this is correct. Regions should form a lattice with `Global` on the top and an empty (invalid) region on the bottom, and any path through the lattice should contract the region. Why wouldn't this code let short-lived values escape?

Reply:
Agreed – but this only sets the region once. I'm not convinced it is correct either, hence the discussion yesterday, but how (where) else is the region initialised for assignments?

Reply:
(Note the `present` check.)

Reply:
Let's continue the discussion in #1301.